### ISSUE TITLE
Add NVCC_WRAPPER_DEFAULT_ARCH Variable

### ISSUE THREAD

**Issue by lkersting** [kokkos/kokkos#3667]
Currently the nvcc_wrapper sets `default_arch="sm_35"`. It would be useful to have an environment variable (e.g. NVCC_WRAPPER_DEFAULT_ARCH) that could be used to set `default_arch` in nvcc_wrapper. Something similar is done for the host compiler:

~~~cmake
host_compiler=${NVCC_WRAPPER_DEFAULT_COMPILER:-"g++"}
~~~

In my case I have libraries that are compiled by nvcc_wrapper but do not link with Kokkos (i.e. do not get the compile flags set by Kokkos).  I set the proper `Kokkos_ENABLE_ARCH_###=ON` in CMake but I also have to add the `-arch=sm_##` compile flag to every package/library that doesn't link with Kokkos.

**Comment by crtrott** [comment]
sounds reasonable. @jrmadsen can you take a stab at that? Essentially just mirror what the NVCC_WRAPPER_DEFAULT_COMPILER does?

**Comment by jrmadsen** [comment]
Yeah, I'll take care of it.

@lkersting is this necessary bc these libraries are using raw CUDA? If not, you could probably just use the compiler launcher feature so they won't get compiled by nvcc_wrapper (to enable this, just don't set the CXX compiler to nvcc_wrapper)

**Comment by lkersting** [comment]
They don't use raw CUDA, but once you set the CXX compiler for a project it can't be changed. For several TPL I have been using `target_compiler_option( ${PKG} PRIVATE  --host_only)` but I'm still getting the following warning for internal dependencies of these TPLs :

> nvcc warning : The 'compute_35', 'compute_37', 'compute_50', 'sm_35', 'sm_37' and 'sm_50' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).

To solve it I have been passing the INTERFACE_LINK_OPTIONS and INTERFACE_COMPILE_OPTIONS from KokkosCore to these packages.

**Comment by jrmadsen** [comment]
> once you set the CXX compiler for a project it can't be changed

Yes, I am saying _do not_ set the CXX compiler to `nvcc_wrapper` for the entire project. Basically, imagine you knew absolutely nothing about `nvcc_wrapper` because you didn't read the docs and you just blindly configured kokkos with `-DKokkos_ENABLE_CUDA=ON`.

The launch-compiler feature effectively _does_ allow you to have two different "C++" compilers -- KokkosCore adds `-DKOKKOS_DEPENDENCE` to its public compile and link options, then using a cmake feature, all compile and link commands are re-routed through the `kokkos_launch_compiler` script, this script looks for the `-DKOKKOS_DEPENDENCE` flag and if it finds it, replaces `${CMAKE_CXX_COMPILER}` with `nvcc_wrapper`. If that flag is not found, it just executes the compile or link command as-is. Thus:

> To solve it I have been passing the INTERFACE_LINK_OPTIONS and INTERFACE_COMPILE_OPTIONS from KokkosCore (when linking to an installed version of Kokkos) or KOKKOS_LINK_OPTIONS and KOKKOS_CUDA_OPTIONS (when including Kokkos as a subfolder) to these packages.

You can get rid of all this, none of the TPLs will be compiled or linked via `nvcc_wrapper` unless they depend on a Kokkos target (which will then add `-DKOKKOS_DEPENDENCE` to its flags).

**Comment by jrmadsen** [comment]
I'll still get this featured added but I expect the launch-compiler method is a much, much better solution. You can also do:

```cmake
find_package(Kokkos COMPONENTS separable_compilation)
```

and then apply the launch-compiler rule per CMake project, directory, target, or source. E.g.:

```cmake
# set all compile and link commands to be re-routed through kokkos_launch_compiler
kokkos_compilation(GLOBAL)

# apply the rule to all targets in a cmake project
kokkos_compilation(PROJECT)

# apply the rule to specific directories/targets/source-files
kokkos_compilation(
  DIRECTORY <list-of-directories>
  TARGET <list-of-targets>
  SOURCE <list-of-source-files>
)
```

**Comment by lkersting** [comment]
Thanks, I didn't realized Kokkos had this launch-compiler feature, when was it added?
Also, what are the differences between:
`target_link_libraries(${PKG} PRIVATE Kokkos::kokkos)`
and
`kokkos_compilation(TARGET ${PKG} )`

Does the second just add the `-DKOKKOS_DEPENDENCE` flag?

**Comment by lkersting** [comment]
Maybe I'm still not understanding this correctly. I have my project which uses the normal gcc compiler then in CMake I have:

~~~cmake
set(Kokkos_ENABLE_CUDA ON CACHE BOOL "" FORCE)
add_subdirectory(${KOKKOS_DIR})
~~~

but I'm get the following error:
~~~
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found CUDA: /projects/sems/install/rhel7-x86_64/sems/compiler/cuda/11.1/base (found version "11.1") 
-- Setting policy CMP0074 to use <Package>_ROOT variables
-- The project name is: Kokkos
-- SERIAL backend is being turned on to ensure there is at least one Host space. To change this, you must enable another host execution space and configure with -DKokkos_ENABLE_SERIAL=OFF or change CMakeCache.txt
-- Using -std=c++17 for C++17 standard as feature
CMake Error at tpl/kokkos/cmake/kokkos_test_cxx_std.cmake:120 (MESSAGE):
  Invalid compiler for CUDA.  The compiler must be nvcc_wrapper or Clang, but
  compiler ID was GNU
Call Stack (most recent call first):
  tpl/kokkos/cmake/kokkos_tribits.cmake:201 (INCLUDE)
  tpl/kokkos/CMakeLists.txt:166 (KOKKOS_SETUP_BUILD_ENVIRONMENT)
~~~

Do I need to set some other variable?

**Comment by jrmadsen** [comment]
It looks like you have too old of a version in `tpl/kokkos`. The error message should read (if you were to get to it, which you shouldn't):

```
"Invalid compiler for CUDA.  The compiler must be nvcc_wrapper or Clang or use kokkos_launch_compiler, but compiler ID was ${KOKKOS_CXX_COMPILER_ID}"
```

> when was it added (I see it is in the develop branch but not the master branch)?

Late September in #3136

**Comment by jrmadsen** [comment]
> Also, what are the differences between:
> `target_link_libraries(${PKG} PRIVATE Kokkos::kokkos)`
> and
> `kokkos_compilation(TARGET ${PKG} )`

Basically, nothing unless you did the `find_package(Kokkos COMPONENTS separable_compilation)`. The `kokkos_compilation` macro does not explicitly add `-DKOKKOS_DEPENDENCE`, it just sets a property on the target to use the compiler launcher, e.g.:

```cmake
SET_PROPERTY(TARGET ${PKG} PROPERTY RULE_LAUNCH_COMPILE "${Kokkos_COMPILE_LAUNCHER} ${CMAKE_CXX_COMPILER}")
SET_PROPERTY(TARGET ${PKG} PROPERTY RULE_LAUNCH_LINK "${Kokkos_COMPILE_LAUNCHER} ${CMAKE_CXX_COMPILER}")
```

but without the `separable_compilation` in the `find_package` above, and if these 2 conditions are met:

1. `Kokkos_ENABLE_CUDA` is ON
2. Failure of an internal test checking whether the C++ compiler can compile CUDA, i.e.:
    - `clang++` would pass this test
    - `g++` would fail this test
    - `mpicxx` with `MPICH_CXX=nvcc_wrapper` in environment would pass this test

Kokkos will do this:

```cmake
SET_PROPERTY(GLOBAL PROPERTY RULE_LAUNCH_COMPILE "${Kokkos_COMPILE_LAUNCHER} ${CMAKE_CXX_COMPILER}")
SET_PROPERTY(GLOBAL PROPERTY RULE_LAUNCH_LINK "${Kokkos_COMPILE_LAUNCHER} ${CMAKE_CXX_COMPILER}")
```

Setting that property doesn't add/remove any flags. It just changes how cmake writes the compile/link command to the makefile so that when the target `${PKG}` gets compiled/linked, the original command, e.g. `/usr/bin/g++ ... pkg.cpp -o pkg.o`, is transformed into `kokkos_launch_compiler /usr/bin/g++ /usr/bin/g++ pkg.cpp -o pkg.o`. 

The `kokkos_launch_compiler` then basically does:

```bash
# if DKOKKOS_DEPENDENCE not in arguments or command exe (at $2) does not match CMAKE_CXX_COMPILER (at $1)
# then execute original command
if [[ -z "$(echo $@ | grep 'DKOKKOS_DEPENDENCE')" || "${1}" != "${2}" ]]; then
    # drop $1
    shift
    # execute original
    eval $@
else
    # drop $1 and $2
    shift
    shift
    nvcc_wrapper $@
fi
```
